File: deep-research-agent/src/deep_research/repl/context.py
"""Context preparation for session continuation."""

from deep_research.agent.state import ResearchSession


def build_continuation_context(session: ResearchSession) -> str:
    """Build compressed context from previous session for continuation.

    Goal: Keep context under 50k tokens
    Strategy: Include insights + report summary, exclude full worker outputs

    Args:
        session: Previous session to build context from

    Returns:
        Compressed context string
    """
    context_parts = []

    # Original query
    context_parts.append(f"PREVIOUS RESEARCH QUERY: {session.query}")

    # Insights (already compressed)
    if session.insights:
        context_parts.append("\nKEY INSIGHTS FROM PREVIOUS RESEARCH:")
        for i, insight in enumerate(session.insights[:10], 1):  # Max 10 insights
            title = insight.get("title", f"Insight {i}")
            finding = insight.get("finding", "")
            context_parts.append(f"{i}. {title}: {finding}")

    # Report summary (first 2000 chars)
    context_parts.append("\nPREVIOUS REPORT SUMMARY:")
    context_parts.append(session.report[:2000])

    # Workers summary
    context_parts.append(f"\n\nRESOURCES: {len(session.workers)} workers executed")
    context_parts.append(f"SOURCES: {len(session.all_sources)} sources consulted")

    return "\n".join(context_parts)


def build_worker_expansion_context(session: ResearchSession, worker_id: str) -> str:
    """Build context for expanding specific worker's research.

    Includes full worker output for deep continuation.

    Args:
        session: Session containing the worker
        worker_id: Worker ID to expand

    Returns:
        Full worker context string

    Raises:
        ValueError: If worker not found
    """
    worker = next((w for w in session.workers if w.worker_id == worker_id), None)

    if not worker:
        raise ValueError(f"Worker {worker_id} not found in session")

    context = f"""PREVIOUS WORKER RESEARCH:

Objective: {worker.objective}
Status: {worker.status}

FULL OUTPUT:
{worker.final_output}

SOURCES:
{chr(10).join(f'- {src}' for src in worker.sources)}
"""

    return context

File: go-research/internal/session/context.go
package session

import (
	"fmt"
	"strings"
)

// BuildContinuationContext creates context for follow-up research
func BuildContinuationContext(sess *Session) string {
	var ctx strings.Builder

	ctx.WriteString(fmt.Sprintf("Previous query: %s\n\n", sess.Query))

	if sess.Report != "" {
		ctx.WriteString("Previous findings:\n")
		// Truncate report if too long
		report := sess.Report
		if len(report) > 2000 {
			report = report[:2000] + "...[truncated]"
		}
		ctx.WriteString(report)
		ctx.WriteString("\n\n")
	}

	if len(sess.Insights) > 0 {
		ctx.WriteString("Key insights:\n")
		for _, insight := range sess.Insights {
			ctx.WriteString(fmt.Sprintf("- %s: %s\n", insight.Title, insight.Finding))
		}
		ctx.WriteString("\n")
	}

	if len(sess.Sources) > 0 {
		ctx.WriteString("Sources consulted:\n")
		limit := len(sess.Sources)
		if limit > 10 {
			limit = 10
		}
		for _, src := range sess.Sources[:limit] {
			ctx.WriteString(fmt.Sprintf("- %s\n", src))
		}
	}

	return ctx.String()
}

// BuildWorkerContext creates context from a specific worker's output
func BuildWorkerContext(worker WorkerContext) string {
	var ctx strings.Builder

	ctx.WriteString(fmt.Sprintf("Worker objective: %s\n\n", worker.Objective))

	if worker.FinalOutput != "" {
		ctx.WriteString("Worker output:\n")
		output := worker.FinalOutput
		if len(output) > 1500 {
			output = output[:1500] + "...[truncated]"
		}
		ctx.WriteString(output)
		ctx.WriteString("\n\n")
	}

	if len(worker.Sources) > 0 {
		ctx.WriteString("Sources found:\n")
		limit := len(worker.Sources)
		if limit > 5 {
			limit = 5
		}
		for _, src := range worker.Sources[:limit] {
			ctx.WriteString(fmt.Sprintf("- %s\n", src))
		}
	}

	return ctx.String()
}

// BuildSynthesisContext creates context for synthesizing multiple worker outputs
func BuildSynthesisContext(workers []WorkerContext) string {
	var ctx strings.Builder

	for i, w := range workers {
		ctx.WriteString(fmt.Sprintf("### Worker %d: %s\n\n", i+1, w.Objective))
		output := w.FinalOutput
		if len(output) > 1000 {
			output = output[:1000] + "...[truncated]"
		}
		ctx.WriteString(output)
		ctx.WriteString("\n\n")
	}

	return ctx.String()
}
